### PR TITLE
Database-Driven Product Server (Part 2)

### PR DESCRIPTION
**Pull request by phillipwright7**
# Summary

In this pull request, I've refactored my product server handlers to use the product database I've seeded in part 1.

# Description

This product server database hopefully meets all acceptance criteria, as I've tested the different scenarios laid out by my instructor. It includes the use of the `database/sql` package and will write designated HTTP response codes to the response header given the certain scenarios in question.

# Output

#### `GET /products?limit=<numeric>`

```
curl -i -X GET http://localhost:8080/products?limit=0 && printf "\n"
HTTP/1.1 200 OK
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:28:47 GMT
Content-Length: 5

null

curl -i -X GET http://localhost:8080/products?limit=-5 && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:28:47 GMT
Content-Length: 0


curl -i -X GET http://localhost:8080/products?limit=a && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:28:48 GMT
Content-Length: 0


curl -i -X GET http://localhost:8080/products?limit=5 && printf "\n"
HTTP/1.1 200 OK
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:28:48 GMT
Content-Length: 325

[{"id":1,"name":"Water - San Pellegrino","description":"","price":80},{"id":2,"name":"Cape Capensis - Fillet","description":"","price":53},{"id":3,"name":"Bread - Bistro Sour","description":"","price":22},{"id":4,"name":"Lamb - Shoulder","description":"","price":16},{"id":5,"name":"iPhone 14","description":"","price":799}]
```

#### `GET /products/<id>`

```
curl -i -X GET http://localhost:8080/products/0 && printf "\n"
HTTP/1.1 404 Not Found
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:28:49 GMT
Content-Length: 0


curl -i -X GET http://localhost:8080/products/-99 && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:28:49 GMT
Content-Length: 0


curl -i -X GET http://localhost:8080/products/b && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:28:49 GMT
Content-Length: 0


curl -i -X GET http://localhost:8080/products/99 && printf "\n"
HTTP/1.1 200 OK
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:28:50 GMT
Content-Length: 76

{"id":99,"name":"Lemonade - Pineapple Passion","description":"","price":20}
```

#### `POST /products`

```
curl -i -X POST http://localhost:8080/products -d '{"price": 2}' && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:02 GMT
Content-Length: 0


curl -i -X POST http://localhost:8080/products -d '{"name": "Apple"}' && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:02 GMT
Content-Length: 0


curl -i -X POST http://localhost:8080/products -d '{"name": "Apple", "price": hfhgdg}' && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:02 GMT
Content-Length: 0


curl -i -X POST http://localhost:8080/products -d '{"name": "Apple", "price": "hfhgdg"}' && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:02 GMT
Content-Length: 0


curl -i -X POST http://localhost:8080/products -d '{"name": "Apple", "price": 2}' && printf "\n"
HTTP/1.1 201 Created
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:02 GMT
Content-Length: 0
```

#### `PUT /products/<id>`

```
curl -i -X PUT http://localhost:8080/products/10000 -d '{"name": "iPhone 14", "price": 799}' && printf "\n"
HTTP/1.1 404 Not Found
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:10 GMT
Content-Length: 0


curl -i -X PUT http://localhost:8080/products/b -d '{"name": "iPhone 14", "price": 799}' && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:10 GMT
Content-Length: 0


curl -i -X PUT http://localhost:8080/products/100 -d '{"name": "iPhone 14"}' && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:10 GMT
Content-Length: 0


curl -i -X PUT http://localhost:8080/products/100 -d '{"price": 799}' && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:10 GMT
Content-Length: 0


curl -i -X PUT http://localhost:8080/products/100 -d '{"name": "iPhone 14", "price": 799}' && printf "\n"
HTTP/1.1 404 Not Found
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:10 GMT
Content-Length: 0
```

#### `DELETE /products/<id>`

```
curl -i -X DELETE http://localhost:8080/products/10000 && printf "\n"
HTTP/1.1 404 Not Found
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:16 GMT
Content-Length: 0


curl -i -X DELETE http://localhost:8080/products/c && printf "\n"
HTTP/1.1 400 Bad Request
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:16 GMT
Content-Length: 0


curl -i -X DELETE http://localhost:8080/products/100 && printf "\n"
HTTP/1.1 404 Not Found
Content-Type: application/json
Date: Wed, 30 Nov 2022 15:29:16 GMT
Content-Length: 0
```

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -17,11 +19,40 @@ type product struct {
...
-var products []product
+var database *sql.DB
 
 func getProductsHandler(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
-	if err := json.NewEncoder(w).Encode(products); err != nil {
+	var payload []product
+	params := r.URL.Query()["limit"]
+
+	rows, err := database.Query("SELECT id, name, price FROM products ORDER BY id LIMIT " + params[0])
```
Comment (jboursiquot) [inline cmd/server/main.go:29]: Don't do that. You're inviting a SQL injection attack. Use the parameterized query syntax. In addition, you need to validate that limit is indeed a number else your SQL syntax could be invalid if non-numeric value is used for the LIMIT clause.

File: cmd/server/main.go [hunk 2]
```diff
@@ -30,18 +61,41 @@ func getProductsHandler(w http.ResponseWriter, r *http.Request) {
 func getProductHandler(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
 	vars := mux.Vars(r)
-	id, err := strconv.Atoi(vars["id"])
+	var payload product
+	rows, err := database.Query("SELECT id, name, price FROM products WHERE id = " + vars["id"])
```
Comment (jboursiquot) [inline cmd/server/main.go:65]: Same problem here. SQL injection risk.

**Review by jboursiquot — CHANGES_REQUESTED** [review]

**Review by phillipwright7 — COMMENTED** [review]
I've fixed the areas you pointed out, and I've also included my updated output in the PR's comment.

File: cmd/server/main.go [hunk 3]
```diff
@@ -17,31 +19,103 @@ type product struct {
...
+		log.Fatal(err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var id int
+		var name string
+		var price float64
+		if err := rows.Scan(&id, &name, &price); err != nil {
+			log.Fatal(err)
```
Comment (jboursiquot) [inline cmd/server/main.go:49]: Don't crash your own server

File: cmd/server/main.go [hunk 4]
```diff
@@ -17,31 +19,103 @@ type product struct {
...
+			ID:    id,
+			Name:  name,
+			Price: int(price),
+		}
+		payload = append(payload, p)
+	}
+
+	err = rows.Err()
+	if err != nil {
+		log.Fatal(err)
```
Comment (jboursiquot) [inline cmd/server/main.go:61]: Don't crash your server

File: cmd/server/main.go [hunk 5]
```diff
@@ -17,31 +19,103 @@ type product struct {
...
-		log.Printf("error converting id to int: %v", err)
+		w.WriteHeader(400)
+		return
 	}
-	p := productFinder(id)
-	if p == nil {
-		log.Printf("product with id %d is not found", id)
+
+	var payload product
+	rows, err := database.Query("SELECT id, name, price FROM products WHERE id = ?", vars["id"])
```
Comment (jboursiquot) [inline cmd/server/main.go:85]: Use `QueryRow`

File: cmd/server/main.go [hunk 6]
```diff
@@ -17,31 +19,103 @@ type product struct {
...
+	if err != nil {
+		w.WriteHeader(400)
 		return
 	}
-	if err := json.NewEncoder(w).Encode(p); err != nil {
-		log.Printf("error encoding product: %v", id)
-		w.WriteHeader(http.StatusInternalServerError)
+	defer rows.Close()
+
+	for rows.Next() {
```
Comment (jboursiquot) [inline cmd/server/main.go:92]: No need for loop if you use `QueryRow`

File: cmd/server/main.go [hunk 7]
```diff
@@ -17,31 +19,103 @@ type product struct {
...
+			return
+		}
+		payload = product{
+			ID:    id,
+			Name:  name,
+			Price: int(price),
+		}
+	}
+
+	if payload.ID == 0 {
```
Comment (jboursiquot) [inline cmd/server/main.go:107]: Use  the QueryRow's returned Row's Scan to check if you get a ErrNoRows to help with this decision instead.

File: cmd/server/main.go [hunk 8]
```diff
@@ -17,31 +19,103 @@ type product struct {
...
+		w.WriteHeader(404)
+		return
+	}
+
+	err = rows.Err()
+	if err != nil {
+		log.Fatal(err)
```
Comment (jboursiquot) [inline cmd/server/main.go:114]: Don't crash

File: cmd/server/main.go [hunk 9]
```diff
@@ -50,95 +124,167 @@ func addProductHandler(w http.ResponseWriter, r *http.Request) {
...
-		w.WriteHeader(http.StatusInternalServerError)
+		w.WriteHeader(400)
 		return
 	}
-	p.ID = len(products) + 1
-	products = append(products, p)
-	if err := json.NewEncoder(w).Encode(products); err != nil {
-		log.Printf("error encoding product: %v", err)
-		w.WriteHeader(http.StatusInternalServerError)
+	if p.Name == "" || p.Price == 0 {
```
Comment (jboursiquot) [inline cmd/server/main.go:130]: Put these in a `validate` method on `Product`

File: cmd/server/main.go [hunk 10]
```diff
@@ -50,95 +124,167 @@ func addProductHandler(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	tx, err := database.Begin()
+	if err != nil {
+		w.WriteHeader(500)
+		return
+	}
+	stmt, err := tx.Prepare("INSERT INTO products (name, price) values(?, ?)")
+	if err != nil {
+		w.WriteHeader(400)
```
Comment (jboursiquot) [inline cmd/server/main.go:142]: 500 might be best here

File: cmd/server/main.go [hunk 11]
```diff
@@ -50,95 +124,167 @@ func addProductHandler(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+	defer stmt.Close()
+
+	_, err = stmt.Exec(p.Name, p.Price)
+	if err != nil {
+		w.WriteHeader(400)
```
Comment (jboursiquot) [inline cmd/server/main.go:149]: 500 here

File: cmd/server/main.go [hunk 12]
```diff
@@ -50,95 +124,167 @@ func addProductHandler(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	if err := tx.Commit(); err != nil {
+		w.WriteHeader(400)
```
Comment (jboursiquot) [inline cmd/server/main.go:154]: 500 here

File: cmd/server/main.go [hunk 13]
```diff
@@ -50,95 +124,167 @@ func addProductHandler(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		w.WriteHeader(400)
 		return
 	}
+
+	var p product
 	if err := json.NewDecoder(r.Body).Decode(&p); err != nil {
-		log.Printf("error decoding product: %v", err)
-		w.WriteHeader(http.StatusInternalServerError)
+		w.WriteHeader(400)
```
Comment (jboursiquot) [inline cmd/server/main.go:177]: 500 here

File: cmd/server/main.go [hunk 14]
```diff
@@ -50,95 +124,167 @@ func addProductHandler(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	tx, err := database.Begin()
+	if err != nil {
+		w.WriteHeader(500)
+		return
+	}
+	stmt, err := tx.Prepare("UPDATE products SET name = ?, price = ? WHERE id = ?")
+	if err != nil {
+		w.WriteHeader(400)
```
Comment (jboursiquot) [inline cmd/server/main.go:192]: 500 here

File: cmd/server/main.go [hunk 15]
```diff
@@ -50,95 +124,167 @@ func addProductHandler(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+	defer stmt.Close()
+
+	result, err := stmt.Exec(p.Name, p.Price, vars["id"])
+	if err != nil {
+		w.WriteHeader(400)
```
Comment (jboursiquot) [inline cmd/server/main.go:199]: 500 here

File: cmd/server/main.go [hunk 16]
```diff
@@ -50,95 +124,167 @@ func addProductHandler(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	if err := tx.Commit(); err != nil {
+		w.WriteHeader(400)
```
Comment (jboursiquot) [inline cmd/server/main.go:204]: 500 here

File: cmd/server/main.go [hunk 17]
```diff
@@ -50,95 +124,167 @@ func addProductHandler(w http.ResponseWriter, r *http.Request) {
...
+
+	rows, err := result.RowsAffected()
+	if err != nil {
+		w.WriteHeader(500)
+		return
+	}
+	if rows == 0 {
+		w.WriteHeader(404)
 		return
 	}
```
Comment (jboursiquot) [inline cmd/server/main.go:216]: Perform a lookup of the product to be updated earlier in the handler.

File: cmd/server/main.go [hunk 18]
```diff
@@ -50,95 +124,167 @@ func addProductHandler(w http.ResponseWriter, r *http.Request) {
...
-}
 
-func productFinder(id int) *product {
-	for _, p := range products {
-		if p.ID == id {
-			return &p
-		}
+	if err := tx.Commit(); err != nil {
+		w.WriteHeader(400)
```
Comment (jboursiquot) [inline cmd/server/main.go:246]: 500 here

File: cmd/server/main.go [hunk 19]
```diff
@@ -50,95 +124,167 @@ func addProductHandler(w http.ResponseWriter, r *http.Request) {
...
+
+	rows, err := result.RowsAffected()
+	if err != nil {
+		w.WriteHeader(500)
+		return
+	}
+	if rows == 0 {
+		w.WriteHeader(404)
+		return
 	}
```
Comment (jboursiquot) [inline cmd/server/main.go:258]: Not necessary. If you want to check the prod exists before deleting, do so earlier in the handler

**Review by jboursiquot — CHANGES_REQUESTED** [review]

**Review by jboursiquot — APPROVED** [review]

**Pull request merged**